docs/source/jit_language_reference_v2.rst

```diff
@@ -0,0 +1,251 @@
+
+.. _language-reference:
+
+TorchScript Language Reference
+==============================
+
+.. _type_annotation:
+
+
+Type Annotation
+~~~~~~~~~~~~~~~
+Since TorchScript is a statically typed, programmers need to annotate types at *strategic points* of TorchScript codes so that every local variable or
+instance data attribute has a static type, and every function and method has a statically typed signature.
+
+When to annotate types
+^^^^^^^^^^^^^^^^^^^^^^
+In general, type annotations are only needed in places where static types cannot be automatically inferred, such as parameters or sometimes return types to
+methods or functions. Types of local variables, data attributes are often automatically inferred from its assignment statements. Sometimes, an inferred type
+may be too restrictive, e.g., ``x`` being inferred as ``NoneType`` through assignment ``x = None``, whereas ``x`` is actually used as an ``Optional``. In such
+cases, type annotations may be needed to overwrite auto inference, e.g., x: ``Optional[int] = None``. Note that it is always safe to type annotate a local variable
+or data attribute even if its type can be automatically inferred. But the annotated type must be congruent with TorchScript’s type checking.
+
+When a parameter, local variable, or data attribute is not type annotated and its type cannot be automatically inferred, TorchScript assumes it to be a
+default type of ``TensorType``.
+
+Annotate function signature
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Since parameter may not be automatically inferred from the body of the function (including both functions and methods), they need to be type annotated,
+otherwise they assume the default type ``TensorType``.
+
+TorchScript supports two styles for method and function signature type annotation:
+
+* **Python3-style** annotates types directly on the signature. As such, it allows individual parameters be left unannotated
+(whose type will be the default type of ``TensorType``) , or the return type be left unannotated (whose type will be automatically inferred).
+
+::
+
+    Python3Annotation := "def" Identifier [ "(" ParamAnnot* ")" ] [ReturnAnnot] ":"
+                         FuncOrMethodBody
+    ParamAnnot := Identifier [ ":" TSType ] ","
+    ReturnAnnot := "->" TSType
+
+Note that using Python3 style, the type of ``self`` is automatically inferred and should not be annotated.
+
+* **Mypy style** annotates types as a comment right below the function/method declaration. In the My-Py style, since parameter names do not appear
+in the annotation, all parameters have to be annotated.
+
+::
+
+    MyPyAnnotation := "# type:" "(" ParamAnnot* ")" [ ReturnAnnot ]
+    ParamAnnot := TSType ","
+    ReturnAnnot := "->" TSType
+
+**Example 1**
+
+In this example, ``a`` is not annotated and assumes the default type of ``TensorType``, ``b`` is annotated as type ``int``, and the return type is not
+annotated and is automatically inferred as type ``TensorType``.
```

Comment:
Might be worth mentioning that the return type is inferred from all exit-like statements like "return" inside function body rather than unconditionally assuming to be "TensorType"